ext/gitsh/src/line_editor.c

```diff
@@ -0,0 +1,1764 @@
+/************************************************
+
+  Based on Ruby 2.3.0's
+  readline.c - GNU Readline module
+
+  $Author$
+  created at: Wed Jan 20 13:59:32 JST 1999
+
+  Copyright (C) 1997-2008  Shugo Maeda
+  Copyright (C) 2008-2013  Kouji Takao
+
+  $Id$
+
+  Contact:
+   - Kouji Takao <kouji dot takao at gmail dot com> (current maintainer)
+
+************************************************/
+
+#ifdef RUBY_EXTCONF_H
+#include RUBY_EXTCONF_H
+#endif
+
+#include "ruby/config.h"
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <readline/readline.h>
+#include <readline/history.h>
+
+#include "ruby.h"
+#include "ruby/io.h"
+#include "ruby/thread.h"
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#ifdef HAVE_SYS_STAT_H
+#include <sys/stat.h>
+#endif
+
+#if !defined(RARRAY_AREF)
+#define RARRAY_AREF(a, i) RARRAY_PTR(a)[i]
+#endif
+
+static VALUE mGitsh;
+static VALUE mReadline;
+
+#define EDIT_LINE_LIBRARY_VERSION "EditLine wrapper"
+#ifndef USE_INSERT_IGNORE_ESCAPE
+# if !defined(HAVE_EDITLINE_READLINE_H) && defined(RL_PROMPT_START_IGNORE) && defined(RL_PROMPT_END_IGNORE)
+#  define USE_INSERT_IGNORE_ESCAPE 1
+# else
+#  define USE_INSERT_IGNORE_ESCAPE 0
+# endif
+#endif
+
+#define COMPLETION_PROC "completion_proc"
+#define COMPLETION_CASE_FOLD "completion_case_fold"
+static ID completion_proc, completion_case_fold;
+#if USE_INSERT_IGNORE_ESCAPE
+static ID id_orig_prompt, id_last_prompt;
+#endif
+#if defined(HAVE_RL_PRE_INPUT_HOOK)
+static ID id_pre_input_hook;
+#endif
+#if defined(HAVE_RL_SPECIAL_PREFIXES)
+static ID id_special_prefixes;
+#endif
+
+#ifndef HAVE_RL_FILENAME_COMPLETION_FUNCTION
+# define rl_filename_completion_function filename_completion_function
+#endif
+#ifndef HAVE_RL_USERNAME_COMPLETION_FUNCTION
+# define rl_username_completion_function username_completion_function
+#endif
+#ifndef HAVE_RL_COMPLETION_MATCHES
+# define rl_completion_matches completion_matches
+#endif
+
+static int (*history_get_offset_func)(int);
+static int (*history_replace_offset_func)(int);
+static int readline_completion_append_character;
+
+static char **readline_attempted_completion_function(const char *text,
+                                                     int start, int end);
+
+#define OutputStringValue(str) do {\
+    SafeStringValue(str);\
+    (str) = rb_str_conv_enc((str), rb_enc_get(str), rb_locale_encoding());\
+} while (0)\
+
+#if !defined(HAVE_RB_OBJ_HIDE)
+#define rb_obj_hide(str) do { RBASIC(str)->klass = 0; } while (0);
+#endif
+
+#if !defined(HAVE_RB_OBJ_REVEAL)
+#define rb_obj_reveal(str, newKlass) do {\
+  RBASIC(str)->klass = newKlass;\
+} while (0);
+#endif
+
+
+/*
+ * Document-class: Readline
```

Comment:
I should probably change the documentation in this file: it refers to the module as `Readline` (it's now `Gitsh::LineEditor`) and makes reference to supporting libedit (which we no longer support).